File: backend/src/models/timeslot.py
from pydantic import BaseModel, Field, ConfigDict, field_validator
from typing import Optional
from datetime import datetime
from ..database.db import TimeSlotStatus


class TimeSlotBase(BaseModel):
    start_at: datetime
    end_at: datetime
    status: TimeSlotStatus

    @field_validator("end_at")
    @classmethod
    def validate_end_after_start(cls, v, info):
        if "start_at" in info.data and v <= info.data["start_at"]:
            raise ValueError("end_at must be after start_at")
        return v


class TimeSlotCreate(TimeSlotBase):
    zone_id: int = Field(..., gt=0)

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "zone_id": 1,
                "start_at": "2024-01-15T09:00:00",
                "end_at": "2024-01-15T12:00:00",
                "status": "active",
            }
        }
    )


class TimeSlotUpdate(BaseModel):
    zone_id: Optional[int] = Field(None, gt=0)
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    status: Optional[TimeSlotStatus] = None

    @field_validator("end_at")
    @classmethod
    def validate_end_after_start(cls, v, info):
        if (
            v is not None
            and "start_at" in info.data
            and info.data["start_at"] is not None
        ):
            if v <= info.data["start_at"]:
                raise ValueError("end_at must be after start_at")
        return v

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "start_at": "2024-01-15T10:00:00",
                "end_at": "2024-01-15T13:00:00",
                "status": "not_active",
            }
        }
    )


class TimeSlotResponse(TimeSlotBase):
    time_id: int
    zone_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: mobile/lib/models/time_slot.dart
class TimeSlot {
  final int timeId;
  final int zoneId;
  final DateTime startAt;
  final DateTime endAt;
  final String status;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  TimeSlot({
    required this.timeId,
    required this.zoneId,
    required this.startAt,
    required this.endAt,
    required this.status,
    this.createdAt,
    this.updatedAt,
  });

  factory TimeSlot.fromJson(Map<String, dynamic> json) {
    return TimeSlot(
      timeId: json['time_id'] ?? json['timeId'],
      zoneId: json['zone_id'] ?? json['zoneId'],
      startAt: DateTime.parse(json['start_at'] ?? json['startAt']),
      endAt: DateTime.parse(json['end_at'] ?? json['endAt']),
      status: json['status'] ?? 'active',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'time_id': timeId,
      'zone_id': zoneId,
      'start_at': startAt.toIso8601String(),
      'end_at': endAt.toIso8601String(),
      'status': status,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}

